### PR TITLE
[sidechannel] add more test cases for checkSignature (mat-1342)

### PR DESCRIPTION
**Pull request by AyushKaul**
@0xAshish can you please take a look if the added tests are wrong, or is it some bug with the checkSignature.
So, I have added 2 test cases 1. for votes> 2/3 2. votes<2/3 the 1. case passes but the 2. also passes which is not the intended behavior.

cc: @jdkanani

### CODE REVIEW

File: contracts/staking/stakeManager/StakeManager.sol [hunk 1]
```diff
@@ -306,8 +306,8 @@ contract StakeManager is IStakeManager, StakeManagerStorage {
         bool acceptDelegation,
         bytes memory signerPubkey
     ) public onlyWhenUnlocked {
-        require(currentValidatorSetSize() < validatorThreshold);
-        require(amount > minDeposit);
+        require(currentValidatorSetSize() < validatorThreshold, "Validator size is less than threshold!");
+        require(amount > minDeposit, "min deposit linit failed!");
```
Comment (jdkanani) [inline contracts/staking/stakeManager/StakeManager.sol:310]:
```suggestion
        require(amount > minDeposit, "min deposit limit failed!");
```

File: contracts/staking/stakeManager/StakeManager.sol [hunk 2]
```diff
@@ -587,7 +587,8 @@ contract StakeManager is IStakeManager, StakeManagerStorage {
         );
         totalRewards = totalRewards.add(_reward);
         require(
-            _stakePower >= currentValidatorSetTotalStake().mul(2).div(3).add(1)
+            _stakePower >= currentValidatorSetTotalStake().mul(2).div(3).add(1),
+            "2/3 non-majority!"
```
Comment (jdkanani) [inline contracts/staking/stakeManager/StakeManager.sol:591]:
```suggestion
            "2/3+1 non-majority!"
```

File: contracts/staking/stakeManager/StakeManager.sol (hunk 1 above)
Comment (AyushKaul) [inline contracts/staking/stakeManager/StakeManager.sol:310]: limit

File: test/staking/StakeManager.test.js [hunk 3]
```diff
@@ -784,6 +785,154 @@ contract('StakeManager: Heimdall fee', async function (accounts) {
...
+      })
+      await stakeManager.stake(stakeAmount, fee, false, wallets[2].getPublicKeyString(), {
+        from:  wallets[2].getAddressString()
+      })
+
+      await stakeToken.mint( wallets[3].getAddressString(), amount)
+      await stakeToken.approve(stakeManager.address, amount, {
+        from:  wallets[3].getAddressString()
+      })
+      await stakeManager.stake(stakeAmount, fee, false, wallets[3].getPublicKeyString(), {
```
Comment (AyushKaul) [inline test/staking/StakeManager.test.js:825]: lot of redundant code can use a loop instead

File: contracts/staking/stakeManager/StakeManager.sol (hunk 1 above)
Comment (AyushKaul) [reply contracts/staking/stakeManager/StakeManager.sol]: done

File: contracts/staking/stakeManager/StakeManager.sol (hunk 2 above)
Comment (AyushKaul) [reply contracts/staking/stakeManager/StakeManager.sol]: made the change

File: test/staking/StakeManager.test.js (hunk 3 above)
Comment (AyushKaul) [reply test/staking/StakeManager.test.js]: done

File: test/staking/StakeManager.test.js [hunk 4]
```diff
@@ -784,6 +785,119 @@ contract('StakeManager: Heimdall fee', async function (accounts) {
...
+        await stakeToken.approve(stakeManager.address, amount, {
+          from: wallet.getAddressString()
+        })
+        await stakeManager.stake(stakeAmount, fee, false, wallet.getPublicKeyString(), {
+          from: wallet.getAddressString()
+        })
+        counter += 1
+      })
+
+      if (counter === 4) {
```
Comment (jdkanani) [inline test/staking/StakeManager.test.js:809]: Not sure why `counter` variable is needed!

File: test/staking/StakeManager.test.js [hunk 5]
```diff
@@ -784,6 +785,119 @@ contract('StakeManager: Heimdall fee', async function (accounts) {
...
```
Comment (jdkanani) [inline test/staking/StakeManager.test.js:863]: Same here!

File: test/staking/StakeManager.test.js [hunk 6]
```diff
@@ -784,6 +785,119 @@ contract('StakeManager: Heimdall fee', async function (accounts) {
...
+
+    it('Less than 2/3rd vote for yes and the rest vote no', async function () {
+      wallets = generateFirstWallets(mnemonics, 4)
+
+      const validators = [1, 2, 3, 4]
+      const amount = web3.utils.toWei('4')
+      let fee = web3.utils.toWei('2')
+      let stakeAmount = web3.utils.toWei('2')
+      
+      let counter = 0
```
Comment (jdkanani) [inline test/staking/StakeManager.test.js:850]: Please remove unwanted counter

File: test/staking/StakeManager.test.js [hunk 7]
```diff
@@ -784,6 +785,119 @@ contract('StakeManager: Heimdall fee', async function (accounts) {
...
     })
+
+    it('More than 2/3rd vote for yes and the rest vote no', async function () {
+      wallets = generateFirstWallets(mnemonics, 4)
+
+      const validators = [1, 2, 3, 4]
+      const amount = web3.utils.toWei('4')
+      let fee = web3.utils.toWei('2')
+      let stakeAmount = web3.utils.toWei('2')
+      let counter = 0
```
Comment (jdkanani) [inline test/staking/StakeManager.test.js:796]: Remove unwanted counter

File: test/staking/StakeManager.test.js (hunk 4 above)
Comment (AyushKaul) [reply test/staking/StakeManager.test.js:809]: to make the code synchronous

Comment (AyushKaul) [reply test/staking/StakeManager.test.js]: done

File: test/staking/StakeManager.test.js [hunk 8]
```diff
@@ -784,6 +785,108 @@ contract('StakeManager: Heimdall fee', async function (accounts) {
...
+        })
+      }
+
+      fee = web3.utils.toWei('1')
+      accountState[1] = [web3.utils.toHex(fee.toString()), 0]
+      accountState[2] = [0, 0]
+      accountState[3] = [0, 0]
+      accountState[4] = [0, 0]
+      // validatorId, accumBalance, accumSlashedAmount, amount
+      let tree = await rewradsTreeFee(validators, accountState)
```
Comment (AyushKaul) [inline test/staking/StakeManager.test.js:814]: this one @0xAshish

**Pull request merged**